packages/components/src/button/test/index.js

```diff
@@ -12,172 +13,197 @@ import { plusCircle } from '@wordpress/icons';
 /**
  * Internal dependencies
  */
-import ButtonWithForwardedRef, { Button } from '../';
-import { VisuallyHidden } from '../../visually-hidden';
+import Button from '../';
+
+jest.mock( '../../icon', () => () => <div data-testid="test-icon" /> );
+jest.mock( '../../tooltip', () => ( { text, children } ) => (
+	<div data-testid="test-tooltip" title={ text }>
+		{ children }
+	</div>
+) );
+jest.mock( '../../visually-hidden', () => ( {
+	__esModule: true,
+	VisuallyHidden: ( { children } ) => (
+		<div data-testid="test-visually-hidden">{ children }</div>
+	),
+} ) );
 
 describe( 'Button', () => {
 	describe( 'basic rendering', () => {
 		it( 'should render a button element with only one class', () => {
-			const button = shallow( <Button /> ).find( 'button' );
-			expect( button.hasClass( 'components-button' ) ).toBe( true );
-			expect( button.hasClass( 'is-large' ) ).toBe( false );
-			expect( button.hasClass( 'is-primary' ) ).toBe( false );
-			expect( button.hasClass( 'is-pressed' ) ).toBe( false );
-			expect( button.prop( 'disabled' ) ).toBeUndefined();
-			expect( button.prop( 'aria-disabled' ) ).toBeUndefined();
-			expect( button.prop( 'type' ) ).toBe( 'button' );
-			expect( button.type() ).toBe( 'button' );
+			const button = render( <Button /> ).container.firstChild;
+
+			expect( button ).toHaveClass( 'components-button' );
+			expect( button ).not.toHaveClass( 'is-large' );
+			expect( button ).not.toHaveClass( 'is-primary' );
+			expect( button ).not.toHaveClass( 'is-pressed' );
+			expect( button ).not.toHaveAttribute( 'disabled' );
+			expect( button ).not.toHaveAttribute( 'aria-disabled' );
+			expect( button ).toHaveAttribute( 'type', 'button' );
+			expect( button.tagName ).toBe( 'BUTTON' );
```

Comment:
I think something like this would be more in line with how we want to write tests in the components package:

```suggestion
			render( <Button /> );
			const button = screen.getByRole( 'button' );

			expect( button ).toHaveClass( 'components-button' );
			expect( button ).not.toHaveClass( 'is-large' );
			expect( button ).not.toHaveClass( 'is-primary' );
			expect( button ).not.toHaveClass( 'is-pressed' );
			expect( button ).not.toBeDisabled();
			expect( button ).toHaveAttribute( 'type', 'button' );
```

The differences being:

1. We try not to use `container.firstChild`. This follows [official RTL guidance](https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#not-using-screen). In this case, it isn't guaranteed that the first child will always be the `button` (could be `div` or whatever).
2. `tagName` assertions can usually be covered with the `screen` query.
3. Using a more abstracted jest-dom matcher (`toBeDisabled()`) can often be more robust than trying to check specific attributes (`disabled`/`aria-disabled`). (It might be that some of the tests in this file do actually need to check specific aria attributes, and that's ok.)

Reply:
> We try not to use `container.firstChild`. This follows [official RTL guidance](https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#not-using-screen). In this case, it isn't guaranteed that the first child will always be the button (could be div or whatever).

I might be misunderstanding, but I think that rule is talking about something else. The rule you're linking is talking about how we run `get` and `query` queries, namely, it suggests that we always call them from `screen`:

```
// ❌
const {getByRole} = render(<Example />)
const errorMessageNode = getByRole('alert')

// ✅
render(<Example />)
const errorMessageNode = screen.getByRole('alert')
```

This doesn't have anything to do with `container.firstChild`. I actually think that for our test it's valuable to assert that the root element is the button, and if that changes, the test should break. 

There is [this "common mistake"](https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#using-container-to-query-for-elements) that is closer to what we're doing here:

```
// ❌
const {container} = render(<Example />)
const button = container.querySelector('.btn-primary')
expect(button).toHaveTextContent(/click me/i)

// ✅
render(<Example />)
screen.getByRole('button', {name: /click me/i})
```

but it talks specifically against using `container.querySelector()` and `container.querySelectorAll()` which we are consciously avoiding here anyway.

The only reason to avoid this would be if we followed the `testing-library/no-node-access` ESLint rule, which we currently aren't. And even if we did, this rule does have a `allowContainerFirstChild` option that allows us to use `container.firstChild` if we want to. 

Sorry for the wall of text 😉 Just wanted to share some of why I've been using that and I've been finding it useful. Sometimes it's just predictable to access the root element of the component, without having to seek a special way to locate it with a query. 

I'm happy to always use `screen` queries, but then, this has its downsides sometimes, especially in components where we have nested elements that match the top elements, or recursive structures. 
Just wanted to hear more of what you think about all this.

> tagName assertions can usually be covered with the screen query.

I'm not sure that's always the case, though - `screen.getByRole( 'button' )` will also find `input[type='button']`, which is definitely unwanted. The `tagName` assertion seems more precise to me in this scenario. WDYT?

> Using a more abstracted jest-dom matcher (toBeDisabled()) can often be more robust than trying to check specific attributes

Are you saying that `toBeDisabled()` is enough for us and is preferred to testing both the `disabled` and `aria-disabled` attributes separately?

Reply:
> I might be misunderstanding, but I think that rule is talking about something else.

Ah sorry, I was referring to this part:

> The only exception to this is if you're setting the `container` or `baseElement` which you probably should avoid doing (I honestly can't think of a legitimate use case for those options anymore and they only exist for historical reasons at this point).

---

> I actually think that for our test it's valuable to assert that the root element is the button, and if that changes, the test should break.

The question is — why should the test break if the root element is a div? Would it break behavior? No. Would it break styling? Maybe. So if we're trying to test for styling breakage with RTL, we're kind of using the wrong tool. Does it warrant a DOM snapshot then? I'd say no in this case.

Even stuff like `.toHaveClass( 'is-foo' )` is not something we generally do, since it's testing an implementation detail. Ideally, it's something for a visual regression test.

> I'm not sure that's always the case, though - `screen.getByRole( 'button' )` will also find `input[type='button']`, which is definitely unwanted. The `tagName` assertion seems more precise to me in this scenario. WDYT?

We usually have no reason to distinguish between implementations as long as they are accessible, and this is what I understand to be the central ethos of RTL. It's why the first choice of query is `getByRole`, and there is no `getByTagName` query or `toHaveTagName` assertion. So if there is a concrete reason that it has to be a `tagName=BUTTON`, I would first try to assert for that behavior (e.g. can I pass HTML as children?), and if that isn't possible, add a comment on why it needs to be a `<button>`.

> Are you saying that `toBeDisabled()` is enough for us and is preferred to testing both the `disabled` and `aria-disabled` attributes separately?

Apologies, I was under the impression that `toBeDisabled` checks both, but I guess [it doesn't](https://github.com/testing-library/jest-dom/issues/144#issuecomment-577235097)! So in a component like this that needs to specifically check for `aria-disabled` behavior because of the "focusable disabled" stuff, I understand it makes sense to assert separately and specifically.

Reply:
This has been very helpful, thank you ❤️  This helped a lot to clear some confusion in my head and to deeply understand the tradeoffs we're making when using this library.

I've updated the tests to reflect your suggestions - namely, no `container` usage, no DOM traversing and using `screen` queries instead.